File: backend/src/main/java/com/onair/hearit/dto/response/RandomHearitResponse.java
package com.onair.hearit.dto.response;

import com.onair.hearit.domain.Bookmark;
import com.onair.hearit.domain.Hearit;
import com.onair.hearit.domain.Keyword;
import java.util.List;

public record RandomHearitResponse(
        Long id,
        String title,
        String categoryColorCode,
        Boolean isBookmarked,
        Long bookmarkId,
        List<KeywordResponse> keywords
) {
    public static RandomHearitResponse from(Hearit hearit, List<Keyword> keywords) {
        List<KeywordResponse> keywordResponses = getKeywordNames(keywords);
        return new RandomHearitResponse(
                hearit.getId(),
                hearit.getTitle(),
                hearit.getCategory().getColorCode(),
                false,
                null,
                keywordResponses
        );
    }

    public static RandomHearitResponse fromWithBookmark(Hearit hearit, Bookmark bookmark, List<Keyword> keywords) {
        List<KeywordResponse> keywordResponses = getKeywordNames(keywords);
        return new RandomHearitResponse(
                hearit.getId(),
                hearit.getTitle(),
                hearit.getCategory().getColorCode(),
                true,
                bookmark.getId(),
                keywordResponses
        );
    }

    private static List<KeywordResponse> getKeywordNames(List<Keyword> keywords) {
        return keywords.stream()
                .map(KeywordResponse::from)
                .toList();
    }

    private record KeywordResponse(
            Long id,
            String name
    ) {
        public static KeywordResponse from(Keyword keyword) {
            return new KeywordResponse(
                    keyword.getId(),
                    keyword.getName()
            );
        }
    }
}

File: android/app/src/main/java/com/onair/hearit/data/dto/RandomHearitResponse.kt
package com.onair.hearit.data.dto

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class RandomHearitResponse(
    @SerialName("content")
    val content: List<Content>,
    @SerialName("isEmpty")
    val isEmpty: Boolean,
    @SerialName("cursorId")
    val cursorId: Long,
) {
    @Serializable
    data class Content(
        @SerialName("id")
        val id: Long,
        @SerialName("title")
        val title: String,
        @SerialName("categoryColorCode")
        val categoryColorCode: String,
        @SerialName("isBookmarked")
        val isBookmarked: Boolean,
        @SerialName("bookmarkId")
        val bookmarkId: Long?,
        @SerialName("keywords")
        val keywords: List<KeywordResponse>,
    )
}
